### PR TITLE
Pull to Filter for Chat Rooms

### PR DESCRIPTION
**Pull request by gnunicorn**
- [x] Pull-To-Filter UI for Chat list
- [x] Expose `isFavorite` and `isDm` and allow to filter by them
- [x] keep filter between restarts
- [x] allow user to favorite chat rooms
- [x] use rust for permalink generation for chat rooms
- [x] add Changelog  

See it in action: The Chat is bookmarked and then we go through the filters:

https://github.com/acterglobal/a3/assets/40496/f7b0f7c7-686c-4884-9e64-6a3d51e8c9f4

And once you scroll, it closes, but shows the "filtered chats" and a filter-with-badge icon in the top bar:

https://github.com/acterglobal/a3/assets/40496/078f0245-c2fb-4823-a2df-0ba4a6b0fbbd

### CODE REVIEW

File: app/lib/common/providers/notifiers/client_pref_notifier.dart [hunk 1]
```diff
@@ -0,0 +1,218 @@
+/// Based on https://github.com/gamako/shared_preferences_riverpod/blob/master/lib/shared_preferences_riverpod.dart
```
Comment (kumarpalsinh25) [inline app/lib/common/providers/notifiers/client_pref_notifier.dart:1]: Skipping review of this file as I'm not getting much understanding in this file changes.
`app/lib/common/providers/notifiers/client_pref_notifier.dart`

File: app/lib/features/chat/pages/room_profile_page.dart [hunk 2]
```diff
@@ -31,10 +31,169 @@ class RoomProfilePage extends ConsumerWidget {
...
+      extendBodyBehindAppBar: true,
+      body: SingleChildScrollView(
+        child: Column(
+          children: [
+            header(context, ref),
+            actions(context, ref),
+            description(context, ref),
+            optionsBody(context, ref),
+          ],
+        ),
```
Comment (kumarpalsinh25) [inline app/lib/features/chat/pages/room_profile_page.dart:45]: 👏

File: app/lib/features/chat/widgets/rooms_list.dart [hunk 3]
```diff
@@ -11,25 +13,174 @@ import 'package:flutter_svg/flutter_svg.dart';
...
+        AppLocalizations.of(context)!.chat,
+        style: Theme.of(context).textTheme.headlineSmall,
+      );
+    }
+
+    return Text(
+      'Filtered chats',
+      style: Theme.of(context).textTheme.headlineSmall,
+    );
+  }
```
Comment (kumarpalsinh25) [inline app/lib/features/chat/widgets/rooms_list.dart:83]:
```suggestion
 Widget roomListTitle(BuildContext context) {
    String title = AppLocalizations.of(context)!.chat;

    if (ref.watch(hasRoomFilters)) {
     title = 'Filter chats';
    }

    return Text(
      title,
      style: Theme.of(context).textTheme.headlineSmall,
    );
  }
```

File: app/lib/features/chat/widgets/rooms_list.dart [hunk 4]
```diff
@@ -11,25 +13,174 @@ import 'package:flutter_svg/flutter_svg.dart';
...
+            .watch(roomListFilterProvider.select((value) => value.searchTerm))
+            ?.isNotEmpty ==
+        true;
+    final hasFilters = ref.watch(hasRoomFilters);
+    return Column(
+      mainAxisAlignment: MainAxisAlignment.start,
+      crossAxisAlignment: CrossAxisAlignment.start,
+      children: [
+        const SizedBox(height: 5),
+        SearchBar(
```
Comment (kumarpalsinh25) [inline app/lib/features/chat/widgets/rooms_list.dart:98]: `SearchBar` is the default flutter frame widget for material component library?

File: app/lib/features/chat/widgets/rooms_list.dart [hunk 5]
```diff
@@ -11,25 +13,174 @@ import 'package:flutter_svg/flutter_svg.dart';
...
+              OutlinedButton(
+                onPressed: () {
+                  setState(() {
+                    _isSearchVisible = false;
+                  });
+                },
+                child: const Text('close'),
+              ),
+          ],
+        ),
```
Comment (kumarpalsinh25) [inline app/lib/features/chat/widgets/rooms_list.dart:168]: This `Clear / Close` UX Flow can be improved. Also on UI side, there are a lot spacing on right side of the button.

Screenshot 2024-03-04 at 1 57 09 PM

**Review by kumarpalsinh25 — APPROVED** [review]
Nice improvement in Chat List!
Overall looks good to me. Just aded some minor suggestions :)

**Comment by kumarpalsinh25** [comment]
> And once you scroll, it closes, but shows the "filtered chats" and a filter-with-badge icon in the top bar:
> Uploading down-to-close.mp4…

@gnunicorn 
Looks like second video is not uploaded not he PR Description.

File: .changes/1471-pull-to-filter-chats.md [hunk 6]
```diff
@@ -0,0 +1,2 @@
+- You can now bookmark chat rooms (via the chat room details page)
+- An easier way to search & filter chats: just pull down in the chat rooms list and the filter bar appears. Next to a cleaned up SearchField it also features quick-toggles for bookmarked chats and DMs only. The selection made here will be persisted between app restart for convenience;
```
Comment (gtalha07) [inline .changes/1471-pull-to-filter-chats.md:2]: Suggestion: Could replace ```SearchField``` with chat search for easier understanding :)

File: app/lib/common/providers/notifiers/client_pref_notifier.dart [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+  set state(T value) {
+    assert(false,
+        "Don't use the setter for state. Instead use `await update(value)`.",);
+    Future(() async {
+      await update(value);
+    });
+  }
+}
+
+/// Returns the [Provider] that has access to the value of preferences.
```
Comment (gtalha07) [inline app/lib/common/providers/notifiers/client_pref_notifier.dart:63]: A lot of commented code here. We can remove this!

File: app/lib/common/providers/notifiers/client_pref_notifier.dart (hunk 1 above)
Comment (gtalha07) [reply app/lib/common/providers/notifiers/client_pref_notifier.dart:1]: It seems like this part is preserving client in application state. @gnunicorn Could you elaborate why this change was needed ? Or what improvement it brings?

Comment (gtalha07) [reply app/lib/common/providers/notifiers/client_pref_notifier.dart:1]: It'd be nice to add developer comments here so everybody knows its purpose ;)

Comment (gnunicorn) [reply app/lib/common/providers/notifiers/client_pref_notifier.dart:1]: This is allowing any state to be saved in the local preferences for persistence. The changes made are to reuse out `sharedPref()`-helper and to have the key be bound to the client-id currently selected.

In this PR is used to keep the state of the "chat filter" between restarts of the app.

I'll add some code comments in the file to clarify what this is about.

File: app/lib/features/chat/widgets/rooms_list.dart (hunk 5 above)
Comment (gnunicorn) [reply app/lib/features/chat/widgets/rooms_list.dart:168]: > This `Clear / Close` UX Flow can be improved.

How would you like it to be improved?

File: app/lib/common/providers/notifiers/client_pref_notifier.dart (hunk 7 above)
Comment (gnunicorn) [reply app/lib/common/providers/notifiers/client_pref_notifier.dart:63]: These are doc-comments. The code is example code for the generated docs. It comes from the copy-paste I made.

File: app/lib/features/chat/widgets/rooms_list.dart (hunk 3 above)
Comment (gnunicorn) [reply app/lib/features/chat/widgets/rooms_list.dart]: also added a feature to adapt the title depending on the filter and search value (see comment below).

File: app/lib/features/chat/widgets/rooms_list.dart [hunk 8]
```diff
@@ -11,25 +13,185 @@ import 'package:flutter_svg/flutter_svg.dart';
...
+      });
+    }
+  }
+
+  Widget roomListTitle(BuildContext context) {
+    String title = AppLocalizations.of(context)!.chat;
+
+    if (ref.watch(hasRoomFilters)) {
+      String subFiltered = 'chats';
+      switch (ref
```
Comment (gtalha07) [inline app/lib/features/chat/widgets/rooms_list.dart:69]: keep conditionals simpler for better code reading ;)
```suggestion
    final selectionValue = ref.watch(roomListFilterProvider.select((value) => value.selection)));
      switch (selectionValue){
```

**Review by gtalha07 — APPROVED** [review]
Overall looks good, nice addition for chats!

**Pull request merged**